webfrontend/src/components/setup/setup-tags.tsx

```diff
@@ -0,0 +1,212 @@
+            tags: toBeSubmitted.map(tag => tag.name)
+        }))
+            .then(response => {
+                setIsLoading(false);
+                if (response.status !== 200) {
+                    triggerErrorMessage();
+                } else {
+                    setTags([...tags].concat(toBeSubmitted));
+                    setToBeSubmitted([]);
+                    setShowAlert(false);
+                }
+            });
+    }
+  }, [isLoading, tags, toBeSubmitted]);
+  const searchTag = useCallback((event: React.SyntheticEvent<EventTarget>) => {
+    event.preventDefault();
+    fetch("/profile/search_tags?tag=" + searchParam)
+        .then(response => {
+            if (response.status !== 200) {
+                triggerErrorMessage();
+            } else {
+                response.json().then(response => {
+                    setSearchResults(response.tags.filter((tag:Tag) => toBeSubmitted.find(addedTag => addedTag.name === tag.name) === undefined));
+                    setSearchDone(true);
+                })
+            }
+        })
+  },[searchParam, toBeSubmitted]);
+  const addNewTag = useCallback((index: number) => {
+    if (tags.length + toBeSubmitted.length < tagCountLimit) {
+        setSearchResults(searchResults.filter((_, i) => i !== index));
+        setToBeSubmitted([...toBeSubmitted, searchResults[index]]);
+    } else {
+        addTagMessageButton.current?.click();
+    }
+}, [searchResults, tagCountLimit, tags.length, toBeSubmitted]);
+
+const removeNewTag = useCallback((index: number) => {
+    setToBeSubmitted(toBeSubmitted.filter((_, i) => i !== index));
+  }, [toBeSubmitted]);
+
+const removeTag = useCallback(() => {
+    fetch('/profile/remove_tag', postRequestContent({
+        tag: tagToBeRemoved?.name
+    })).then(response => {
+        if (response.status !== 200) {
+            triggerErrorMessage();
+            return;
+        }
+        setTags(tags.filter(tag => tag !== tagToBeRemoved));
+        setShowRemoveAlert(false);
+        setCanRemoveTag(false);
+    })
+  }, [tagToBeRemoved, tags]);
+
+  return (
+    <div className="add-tag-container">
+        <div className="current-tag-section">
+           <div className="add-tag-title">Your current tags</div>
+           <div className="add-tag-section-body">
+                {tags.map((tag: Tag, index: number) => (
+                    <div key = {index} className="old-tag-div">
+                        <div className="tag-button btn btn-outline-info">
+                            <img src={tag.icon} alt="tag-icon" />
+                            <div>{tag.name}</div>
+                        </div>
+                        {canRemoveTag && <button type="button" className="btn-close" aria-label="Close" onClick={() => {
+                            setTagToBeRemoved(tag);
+                            setShowRemoveAlert(true);
+                            setShowAlert(false);
+                        }}></button>}
```

Comment:
Abstract out complicated callbacks with `useCallback` instead of putting the whole thing here.